ouroboros-consensus-cardano/tools/db-converter/Main.hs

```diff
@@ -0,0 +1,326 @@
+      protocolMagicId
+  where
+    markerPath :: String
+    markerPath = dbPath FilePath.</> Text.unpack dbMarkerFile
+
+validate :: forall blk. (HasProtocolInfo blk, Node.RunNode blk, Show (Header blk)) => Args Unwrapped -> Proxy blk ->  IO ()
+validate Validate {..} _ = do
+  dbDir' <- parseAbsDir =<< canonicalizePath dbDir
+  protocolInfo <- mkProtocolInfo @blk configFile genesisHash requiresNetworkMagic
+  validateChainDb dbDir' protocolInfo onlyImmDB verbose
+validate _ _ = error "impossible"
+
+validateChainDb
+  :: (Node.RunNode blk, Show (Header blk))
+  => Path Abs Dir -- ^ DB directory
+  -> ProtocolInfo IO blk
+  -> Bool -- Immutable DB only?
+  -> Bool -- Verbose
+  -> IO ()
+validateChainDb dbDir protocolInfo onlyImmDB verbose =
+    withRegistry $ \registry -> do
+      let chainDbArgs = mkChainDbArgs registry InFuture.dontCheck
+          (immDbArgs, _, _, _) = fromChainDbArgs chainDbArgs
+      if onlyImmDB then
+        ImmDB.withImmDB immDbArgs $ \immDB -> do
+          immDbTipPoint <- ImmDB.getPointAtTip immDB
+          putStrLn $ "DB tip: " ++ show immDbTipPoint
+      else
+        ChainDB.withDB chainDbArgs $ \chainDB -> do
+          chainDbTipPoint <- atomically $ ChainDB.getTipPoint chainDB
+          putStrLn $ "DB tip: " ++ show chainDbTipPoint
+  where
+
+    ProtocolInfo { pInfoInitLedger = initLedger, pInfoConfig = cfg } =
+      protocolInfo
+
+    tracer
+      | verbose   = contramap show debugTracer
+      | otherwise = nullTracer
+
+    chunkInfo  = Node.nodeImmDbChunkInfo cfg
+
+    mkChainDbArgs registry btime =
+      let args = Node.mkChainDbArgs tracer registry btime
+            (toFilePath dbDir) cfg initLedger chunkInfo
+      in args {
+          ChainDB.cdbImmValidation = ImmDB.ValidateAllChunks
+        }
+
+class HasProtocolInfo blk where
+  mkProtocolInfo :: [FilePath]
+                 -> Hash CB.Raw
+                 -> Bool  -- is it mainnet?
+                 -> IO (ProtocolInfo IO blk)
+
+{-------------------------------------------------------------------------------
+  ByronBlock instance
+-------------------------------------------------------------------------------}
+
+instance HasProtocolInfo ByronBlock where
+  mkProtocolInfo [configFile] genesisHash requiresNetworkMagic = do
+    config <- openGenesisByron configFile genesisHash requiresNetworkMagic
+    return $ mkProtocolInfoByron config
+  mkProtocolInfo ls _ _ =
+     error $
+       "a single genesis file is needed for pure Byron. Given " ++ show ls
+
+openGenesisByron :: FilePath -> Hash CB.Raw -> Bool -> IO CC.Genesis.Config
+openGenesisByron configFile genesisHash requiresNetworkMagic = do
+    econfig <-
+      runExceptT $
+        CC.Genesis.mkConfigFromFile
+          (if requiresNetworkMagic then RequiresMagic else RequiresNoMagic)
+          configFile
+          genesisHash
+    case econfig of
+      Left err     -> throwIO $ MkConfigError err
+      Right config -> return config
```

Comment:
If we're using `either (error . show) return` everywhere, we should do it here too for consistency.